overlord/devicestate/devicestate.go

```diff
@@ -475,45 +476,73 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 }
 
 type remodelVariant struct {
-	localSnapsRequired bool
+	offline        bool
+	localSnaps     []*snap.SideInfo
+	localSnapPaths []string
 }
 
 type pathSideInfo struct {
 	localSi *snap.SideInfo
 	path    string
 }
 
-func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State,
-	pathSI *pathSideInfo, snapName string, opts *snapstate.RevisionOptions,
+func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State, snapName string, snapID string, opts *snapstate.RevisionOptions,
 	userID int, snapStateFlags snapstate.Flags, tracker snapstate.PrereqTracker,
 	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s track changed", snapName)
 	if opts == nil {
 		opts = &snapstate.RevisionOptions{}
 	}
-	if ro.localSnapsRequired {
-		if pathSI == nil {
-			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+	if ro.offline {
+		pathSI := sideInfoAndPathFromID(ro.localSnaps, ro.localSnapPaths, snapID)
+		if pathSI != nil {
+			return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
+				userID, snapStateFlags, tracker, deviceCtx, fromChange)
 		}
-		return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
-			userID, snapStateFlags, tracker, deviceCtx, fromChange)
+
+		info, err := snapstate.CurrentInfo(st, snapName)
```

Comment:
shouldn't we possibly consider also the other revisions on the device and do a revison refresh?  suppose this could be a follow up

Reply:
Could you elaborate on what you mean here? Other revisions of the same snap? Like, looking for different instances?

Reply:
no, I mean we keep potentialy up to 3 revisions for each snap, they are previous revisions that we refreshed away from but the snap is still on the device. They are listed in SnapState.Sequence

Reply:
Okay, I see. Yes, I'll start working on a follow-up PR for that.